File: minigit-3/ts/src/commands/rev-parse.ts
// rev-parse command - Resolve revisions

import { findRepoRoot } from '../utils';
import { resolveRef } from '../refs';

export function revParse(args: string[]): void {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a git repository');
    process.exit(1);
  }

  if (args.length === 0) {
    console.error('fatal: revision required');
    process.exit(1);
  }

  const revision = args[0];
  const sha = resolveRef(revision, repoRoot);

  if (!sha) {
    console.error(`fatal: bad revision '${revision}'`);
    process.exit(1);
  }

  console.log(sha);
}

File: minigit-3/py/commands/rev_parse.py
"""rev-parse command - Resolve revisions."""

import sys
from pathlib import Path


def run(args: list[str]) -> int:
    """Resolve revision to SHA."""
    from utils import find_repo_root
    from refs import resolve_ref

    repo_root = find_repo_root()
    if repo_root is None:
        print("error: not a minigit repository", file=sys.stderr)
        return 1

    if not args:
        print("error: revision required", file=sys.stderr)
        return 1

    target = args[0]

    sha = resolve_ref(repo_root, target)
    if sha is None:
        print(f"error: unknown revision '{target}'", file=sys.stderr)
        return 1

    print(sha)
    return 0
